Feature/autocomplete

tomdng opened the pull request:
@henrievjen made the text editor component as well as the suggestion feature.

I just redid the CSS for the project for the new design and added the navbar in.

Maybe in the future we'll actually make this autocomplete feature but I'm concerned with the pace of the project so I'd rather get the essential stuff pumped out. This includes getting the tutorial display component working (that shows the step-by-step instructions) and getting the drone simulator working.

I'll also deal with the merge conflicts.

tomdng committed:
Added really BAREBONES text-editor component

tomdng committed:
Merge branch 'feature/landing-page' of https://github.com/MissouriMRR/propguard into feature/autocomplete to fix issues with eslint configuration.

henrievjen committed:
fixed double landing page css issue

henrievjen committed:
added template to main tutorials page and added basic styling

henrievjen committed:
moved styled components outside components

tomdng committed:
Fixed import issue with TutorialApp component

henrievjen committed:
autocomplete works for one word

henrievjen committed:
Merge branch 'feature/autocomplete' of https://github.com/MissouriMRR/propguard into feature/autocomplete

henrievjen committed:
autocomplete works for multiple words

henrievjen committed:
when click away from text input suggestion goes away

henrievjen committed:
fixed terminal autocomplete bugs

henrievjen committed:
updated theme

tomdng committed:
Updated visuals for the text editing and tutorial components

tomdng committed:
Updated typing for TextEditor component and fixed a few ESlint/prettier conflicts

tomdng committed:
Cleaned up code on TextEditor component

tomdng committed:
Added a few comments to the TextEditor component for the suggestion function

tomdng committed:
Renamed TutorialDisplay component file to match other components

tomdng committed:
Removed Bootstrap entirely and added the navbar in

tomdng committed:
Finally updated the README for the project

tomdng committed:
Updated README description

ClayMav commented on README.md:
```diff
@@ -1,3 +1,100 @@
+
+## What is this?
+
+Propguard is a webapp built for Missouri S&T's Multirotor Drone Design Team meant to preserve the knowledge and experience of its past members. This means that Propguard teaches new members how to start working on the team's custom flight code via a Codecademy-like tutorial interface.
```
```suggestion
Propguard is a web app built for Missouri S&T's Multirotor Robot Design Team meant to preserve the knowledge and experience of its past members. Propguard teaches new members how to start working on the team's flight code via a Codeacademy-like tutorial interface.
```

ClayMav commented on README.md:
```diff
@@ -1,3 +1,100 @@
+
+Propguard is a webapp built for Missouri S&T's Multirotor Drone Design Team meant to preserve the knowledge and experience of its past members. This means that Propguard teaches new members how to start working on the team's custom flight code via a Codecademy-like tutorial interface.
+
+A second part of Propguard (in development) is to create a wiki that will serve as the collective knowledge of all the past members of the Multirotor team. During our flight tests we always run into issues, and some of those issues happen to be things that have been solved in the past. We obviously don't want to keep making the same mistakes.
```
```suggestion
It includes a wiki that will serve as the collective knowledge of all the past members of the team. During our flight tests we always run into issues, and some of those issues happen to be things that have been solved in the past. We obviously don't want to keep making the same mistakes.
```

ClayMav commented on README.md:
````diff
@@ -1,3 +1,100 @@
+2. Install the required packages for the project
+
+```sh
+yarn
````
Be explicit when writing instructions, this works but explicit is better
```suggestion
yarn install
```

ClayMav commented on README.md:
```diff
@@ -1,3 +1,100 @@
+
+## Roadmap and progress
+
+### Winter Break Goals
```
Give dates

ClayMav commented on README.md:
```diff
@@ -1,3 +1,100 @@
+
+- Finish front-end for the tutorial part of Propguard
+
+### Before Spring Break
```
Give dates

ClayMav commented on README.md:
```diff
@@ -1,3 +1,100 @@
+- Write basic tutorials for setting up development environments for the flight team
+- Create the front-end for the wiki
+
+### Before Finals
```
Give dates

ClayMav commented on README.md:
```diff
@@ -1,3 +1,100 @@
+- Create a backend for the wiki to host content on
+- Come up with a system that lets the wiki by easily editable with some CMS like Ghost+Gatsby or something similar
+
+### Summer 2020 Before Competition
```
Give dates

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -2,91 +2,104 @@ import React from "react";
+      <Main>
+        <Title>Propguard</Title>
+        <Subtext>
+          A webapp that teaches people how to use Dronekit and Python to start
```
```suggestion
          A webapp that teaches people how to use MAVSDK and Python to start
```

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -2,91 +2,104 @@ import React from "react";
+        <Title>Propguard</Title>
+        <Subtext>
+          A webapp that teaches people how to use Dronekit and Python to start
+          and launch the drones
```
```suggestion
          flying copters autonomously
```

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -0,0 +1,164 @@
+    let startIndex = 0;
+
+    // Keeping the type as "any" since DOM stuff isn't super supported
+    const terminal: any = document.getElementById("terminal");
```
Is there a reason we are getting this by id and not just making it

tomdng committed:
Update README.md

Co-Authored-By: Clay McGinnis <mcginnis@claymav.com>

tomdng committed:
Update README.md

Co-Authored-By: Clay McGinnis <mcginnis@claymav.com>

tomdng committed:
Update README.md

Co-Authored-By: Clay McGinnis <mcginnis@claymav.com>

tomdng committed:
README updates and landing page updates

tomdng committed:
Merge branch 'master' into feature/autocomplete

tomdng committed:
Updated readme with new roadmap goals

tomdng committed:
Added a tutorial_data file that'll serve as a data source for the tutorial component

tomdng committed:
Removed uncessary DOM code for the autocomplete function and assigned the right types to everything in it

tomdng commented:
Alright I updated the roadmaps to reflect the stuff that we've talked about @ClayMav and I removed some of the redundant DOM code for the autocomplete feature. Please review the changes I made.

ClayMav commented on .eslintrc.js:
```diff
@@ -27,7 +28,8 @@ module.exports = {
     "react/jsx-filename-extension": ["error", { "extensions": [".js", ".jsx", ".ts", ".tsx"] }],
-    '@typescript-eslint/indent': ['error', 2]
+    '@typescript-eslint/indent': ['error', 2],
+    "react/jsx-curly-newline": 0,
```
kinda looks like the auto-formatter wasn't run on this, since this line has double quotes and the one before is single quotes

ClayMav commented on README.md:
```diff
@@ -1,3 +1,87 @@
+<br />
+<p align="center">
+    <h2 align="center">Propguard</h2>
+    <p align="center">Currently only two software members really know how to start the drone and enable its autonomous features. We need to change that. Propguard is a webapp that teaches people how to use Dronekit and Python to start and launch the drones.</p>
```
On reevaluation, lets ditch the "two software members" and just keep a version of the last sentence. It's too specific and won't be true all the time.

ClayMav commented on README.md:
```diff
@@ -1,3 +1,87 @@
+
+## Usage
+
+We're working on getting a live version up so for now you'll have to run this project locally on your machine. See the **Get Started** section for instructions on how to run it.
```
You use use header links to link down to get started

ClayMav commented on README.md:
```diff
@@ -1,3 +1,87 @@
+- Write a tutorial for writing actual flight code (March 10)
+- Create a backend and database to host propguard tutorial (March 17)
+- Find a place to host the wiki and tutorial (March 19)
+- Come up with a system that lets the wiki by easily editable with some CMS like Ghost+Gatsby or something similar (April 14)
```
Is this still accurate

ClayMav replied:
Not sure that the roadmap should be in the repo anyways, you could do it by using projects and issues in github

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -19,16 +25,18 @@ const Main: AnyStyledComponent = styled.main`
 `;
 
-const Title: AnyStyledComponent = styled.h1`
+const Title: AnyStyledComponent = styled.div`
```
Is there a reason you switched away from a semantic HTML element to a div

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -60,33 +68,39 @@ const Title: AnyStyledComponent = styled.h1`
 
-const Subtext: AnyStyledComponent = styled.aside`
-  font-size: 16px;
+const Subtext: AnyStyledComponent = styled.span`
```
is there a reason to use a non-semantic html element instead of a semantic one

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -3,14 +3,20 @@ import { Link } from "react-router-dom";
-const Main: AnyStyledComponent = styled.main`
-  height: 350px;
-  color: white;
+const Main: AnyStyledComponent = styled.div`
```
if this is truly the main you would think you would use the semantic element

ClayMav replied:
but i don't think this is actually main, so you might consider changing the title

ClayMav commented on src/App.tsx:
```diff
@@ -1,10 +1,13 @@
 import { Landing } from "./components/landing/Landing";
 
-const Main: AnyStyledComponent = styled.main`
+const Main: AnyStyledComponent = styled.div`
```
if this is truly the main you would think you would use the semantic element

ClayMav commented on src/components/navbar/Navbar.tsx:
```diff
@@ -0,0 +1,45 @@
+import { Link } from "react-router-dom";
+import styled, { AnyStyledComponent } from "styled-components";
+
+const NavWrapper: AnyStyledComponent = styled.div`
```
consider adding a "nav" html semantic element somewhere

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -0,0 +1,158 @@
+import React, { useState } from "react";
+import styled, { AnyStyledComponent } from "styled-components";
+
+const Main: AnyStyledComponent = styled.div`
```
maybe stop using the word "Main" for all of your components

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -0,0 +1,158 @@
+  text-align: center;
+`;
+
+const Terminal: AnyStyledComponent = styled.form`
```
does there need to be a form around the text element, since you are controlling it with react it shouldn't be strictly necessary

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -0,0 +1,158 @@
+  const autoCompleteCheck = (
+    event: React.ChangeEvent<HTMLInputElement>
+  ): void => {
+    let text: string = event.target.value;
```
I would prefer you control the react component and have the value in state

ClayMav replied:
https://reactjs.org/docs/forms.html scroll to "The textarea Tag"

ClayMav replied:
Basically just have the onchange be a function that simple updates the state and then do a useeffect hook to check for autocomplete

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -0,0 +1,158 @@
+      if (
+        words[i].startsWith(text) &&
+        text.length !== words[i].length &&
+        text.length !== 0
```
make this the first condition of the if statement

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -0,0 +1,158 @@
+  return (
+    <Main>
+      <Terminal
+        onSubmit={(event: React.FormEvent<HTMLFormElement>): void =>
```
would prefer if it was just `onSubmit={handleSubmit}`

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -0,0 +1,158 @@
+        <Text
+          type="text"
+          onBlur={removeSuggestion}
+          onChange={(event: React.ChangeEvent<HTMLInputElement>): void => {
```
If you control it properly you would just have `onChange={handleChange}` or something like that

ClayMav commented on src/components/tutorial-app/TutorialApp.tsx:
```diff
@@ -0,0 +1,59 @@
+import { TextEditor } from "../text-editor/TextEditor";
+import { TutorialDisplay } from "../tutorial-display/TutorialDisplay";
+
+const TutorialPage: AnyStyledComponent = styled.main`
```
Should this be a main?

ClayMav commented on src/components/tutorial-app/TutorialApp.tsx:
```diff
@@ -0,0 +1,59 @@
+  height: 100%;
+`;
+
+const Parent: AnyStyledComponent = styled.div`
```
Bad name

ClayMav commented on src/components/tutorial-display/TutorialDisplay.tsx:
```diff
@@ -0,0 +1,42 @@
+  width: 80%;
+`;
+
+const TutorialText: AnyStyledComponent = styled.div`
```
Shouldn't this be a semantic element

ClayMav reviewed (changes requested):
Sorry I didn't catch all this earlier. This review was much more in depth than before.

Here is a link about semantic elements: https://www.w3schools.com/html/html5_semantic_elements.asp

tomdng committed:
Turned TextEditor into a controlled component and made HTML more semantic

tomdng committed:
Made TutorialText style in the TutorialDisplay component semantic

tomdng commented:
Alright you definitely made me think about semantic elements which I kinda forget as I was focusing more on pumping out react components.

ClayMav commented on README.md:
```diff
@@ -5,7 +5,7 @@
 <p align="center">
     <h2 align="center">Propguard</h2>
-    <p align="center">Currently only two software members really know how to start the drone and enable its autonomous features. We need to change that. Propguard is a webapp that teaches people how to use Dronekit and Python to start and launch the drones.</p>
+    <p align="center">Propguard is a webapp that teaches people how to start writing basic flight code with technologies like Dronekit, MAVSDK, and Python.</p>
```
```suggestion
    <p align="center">Propguard is a webapp that teaches people how to start writing basic flight code with technologies like PX4, MAVSDK, and Python.</p>
```

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -68,7 +68,7 @@ border-right-color: rgba(0, 0, 0, 1);
 `;
 
-const Subtext: AnyStyledComponent = styled.span`
+const Subtext: AnyStyledComponent = styled.aside`
```
```suggestion
const Subtext: AnyStyledComponent = styled.p`
```

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -14,7 +14,7 @@ const Main: AnyStyledComponent = styled.div`
 `;
 
-const Terminal: AnyStyledComponent = styled.form`
+const Terminal: AnyStyledComponent = styled.div`
```
consider squashing this element

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -65,36 +65,34 @@ const SuggestBox: AnyStyledComponent = styled.div`
+      startIndex = userInput.lastIndexOf(" ");
 
-      if (event.target.value.length <= cursorPos) {
+      if (userInput.length <= cursorPos) {
```
use `text` instead of `userInput`

ClayMav commented on src/components/text-editor/TextEditor.tsx:
```diff
@@ -106,52 +104,44 @@ const TextEditor: React.FC = (): JSX.Element => {
       }
     }
-  };
+  });
```
```suggestion
  }, [userInput, cursorPos]);
```

tomdng committed:
Apply suggestions from code review

Co-Authored-By: Clay McGinnis <mcginnis@claymav.com>

tomdng committed:
Fixed bug where autocomplete wouldn't work on a newline and changed landing page semantic elements
